orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -389,6 +662,7 @@ Status TrainingRunner::TrainingLoop(IDataLoader& training_data_loader, IDataLoad
         }
       }  // end of one file/shard
 
+      pipeline_worker_pool_.join_all();
```

Comment:
>pipeline_worker_pool_.join_all(); [](start = 6, length = 33)

Since pipeline schedule is with fixed gradient accumulation steps, we may need to fill in empty batch if not last step is not gradient accumulation step

Reply:
It could complicate things. Let's add
```
ORT_ENFORCE(num_training_steps % gradient_accumulation_steps == 0)
```
It's pointless if we run extra gradient accumulation without updating the model.

Reply:
If ORT handles batch=0 well, then there should be no extra cost. The proposed ORT_ENFORCE could cause unnecessary trouble for customers with their own data..


---
In reply to: [420466180](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r420466180) [](ancestors = 420466180)